### PR TITLE
Reworked shared basal rate support for Dash & Eros

### PR DESCRIPTION
**Pull request by itsmojo**
+ Dash zero scheduled basal rates now supported
+ Dash zero temp basal corrected for all durations
+ Eros zero temp basal rate clean up work
+ Handle any non-standard basal rates
+ Add new unit test for non-standrad basal rates
+ Add new tests for Dash zero basal rate schedules
+ Add new tests for Dash style zero temp basal
+ Add missing tests from OmniKitTests

### CODE REVIEW

File: OmniBLE/OmnipodCommon/BasalDeliveryTable.swift [hunk 1]
```diff
@@ -159,6 +159,24 @@ extension BasalDeliveryTable: CustomDebugStringConvertible {
     }
 }
 
+// Normalize basal rate by rounding down to pulse size boundary,
+// but basal rates within a small delta will be rounded up.
+func normalizedBasalRate(rate: Double) -> Double
```
Comment (ps2) [inline OmniBLE/OmnipodCommon/BasalDeliveryTable.swift]: I think "rounding" is a better word for this process rather than "normalizing", and would prefer we keep this just a simple rounding that can be explained easily in those few cases we need to. 

Because this should only be used in certain edge cases that we're not currently handling in the UI yet, I don't think it justifies trying to be sophisticated about it. Standard rounding, or next lowest rate are fine.

File: OmniBLE/OmnipodCommon/BasalDeliveryTable.swift [hunk 2]
```diff
@@ -159,6 +159,24 @@ extension BasalDeliveryTable: CustomDebugStringConvertible {
...
+    let delta = 0.01
+    return Pod.supportedBasalRates.last(where: { $0 <= rate + delta }) ?? 0
+}
+
+// Return normalized basal rate for pulse timing purposes.
+// For non-Eros, returns the nearZeroBasalRate (0.01) for a zero basal rate.
+func normalizedBasalTimingRate(rate: Double) -> Double {
+    var nrate = normalizedBasalRate(rate: rate)
+    if nrate == 0.0 {
+        nrate = Pod.zeroBasalRate // will be an adjusted value for non-Eros cases
```
Comment (ps2) [inline OmniBLE/OmnipodCommon/BasalDeliveryTable.swift]: Eros should not have 0 in the supportedBasalRates, so normalizedBasalRate should never return a 0, right?

File: OmniBLE/OmnipodCommon/BasalDeliveryTable.swift [hunk 3]
```diff
@@ -170,58 +188,69 @@ public struct RateEntry {
...
+            return TimeInterval(minutes: 30)
         } else {
-            return round(delayBetweenPulses * Double(totalPulses))
+            // Use delayBetweenPulses to compute duration, works for non-Eros near zero rates
+            return delayBetweenPulses * totalPulses
         }
     }
     
     public var data: Data {
+        var ZZZZZZZZ = UInt32(delayBetweenPulses.hundredthsOfMilliseconds)
```
Comment (ps2) [inline OmniBLE/OmnipodCommon/BasalDeliveryTable.swift]: Please use camel case naming convention for swift variables.

File: OmniBLE/OmnipodCommon/Pod.swift [hunk 4]
```diff
@@ -55,15 +55,28 @@ public struct Pod {
...
     public static let maximumBasalScheduleEntryCount: Int = 24
 
     // Minimum duration of a single basal schedule entry
     public static let minimumBasalScheduleEntryDuration = TimeInterval.minutes(30)
 
+    // Max time between pulses for basal and temp basal extra timing command
+    public static let maxTimeBetweenPulses = TimeInterval(hours: 5)
+
+    // Near zero basal rate used for non-Eros pods for zero scheduled basal rates and temp basals
+    public static let nearZeroBasalRate = 0.01
```
Comment (ps2) [inline OmniBLE/OmnipodCommon/Pod.swift]: I intended the Pod.swift to be more of a class to expose certain user facing constants, not internal protocol constants. Could this be kept more in the encoding part of the code?

**Review by ps2 — COMMENTED** [review]
And nice work finding your way around my protocol encoding layer. It was so touchy, so thanks for bringing over the tests. Generally looks good. Only a few comments on naming and organization.

File: OmniBLE/OmnipodCommon/BasalDeliveryTable.swift (hunk 1 above)
Comment (itsmojo) [reply OmniBLE/OmnipodCommon/BasalDeliveryTable.swift]: I saw values of 0.6499999999999 resulting from 0.65 from some test cases, so this is why I didn't couldn't just round down since this was breaking some test cases. Floating point can be a real pain. I really wish that we did everything in terms of pulses like the PDM does to avoid all these headaches. So then I wanted to avoid using rounding to try to avoid confusion with the similar funcs in the PumpManager files that I couldn't access. Will switch to use "round" instead of "normalize".

File: OmniBLE/OmnipodCommon/BasalDeliveryTable.swift (hunk 3 above)
Comment (itsmojo) [reply OmniBLE/OmnipodCommon/BasalDeliveryTable.swift]: The original name to describe this field was a bit ugly and actually less useful, but I'll change it back.

File: OmniBLE/OmnipodCommon/BasalDeliveryTable.swift (hunk 2 above)
Comment (itsmojo) [reply OmniBLE/OmnipodCommon/BasalDeliveryTable.swift]: "Eros should not have 0 in the supportedBasalRates" -> yes
"so normalizedBasalRate should never return a 0, right?" ->  actually no
Because of the way normalizedBasalRate() was written, it would still return a 0 even if a 0 wasn't in supportedBasalRates[] for Eros (see commenting on OmnipodPumpManager.roundToSupportedBasalRate()).

In any case, this wasn't obvious. Will add explicit comment on the renamed normalizedBasalRate() function stating that it will round down to 0 for both Eros and non-Eros pods and will rewrite this function to not use Pod.supportedBasalRates[] to hopefully make this even more obvious.

File: OmniBLE/OmnipodCommon/Pod.swift (hunk 4 above)
Comment (itsmojo) [reply OmniBLE/OmnipodCommon/Pod.swift]: Moved maxTimeBetweenPulses, nearZeroBasalRate & nearZeroBasalRateFlag declarations to BasalDeliveryTable.swift.
